<a href="https://colab.research.google.com/github/danielamador12/Projetos-Publicos/blob/master/Teste_de_Overfitting_no_classificador_Spotify.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Teste de Overfitting no classificador Spotify
---
[![LinkedIn](https://img.shields.io/badge/LinkedIn-DanielSousaAmador-purple.svg)](https://www.linkedin.com/in/daniel-sousa-amador)
[![GitHub](https://img.shields.io/badge/GitHub-danielamador12-darkblue.svg)](https://github.com/danielamador12)
[![Medium](https://img.shields.io/badge/Medium-DanielSousaAmador-darkorange.svg)](https://medium.com/@daniel.s.amador)


Overfitting é, de uma forma bem direta, um **sobreajuste** do modelo aos dados, ou seja, o modelo de Machine Learning aprende a classificar tanto os modelos de treino que começa a **sofrer para generalizar** os dados de teste. [Aqui](https://github.com/danielamador12/public-projects/blob/master/educacao-comparacao-decisiontree_randomforest.ipynb) você pode ver mais um exemplo de teste de overfitting.

O Overfitting é **muito comum** e normalmente a primeira forma de se conseguir avaliar é, testando com os dados de teste e com os próprios dados de treino.

Vale saber que overfitting é o exato **contraponto do underfitting**, que é o sobajuste, ou seja, quando o modelo fica fraco e não consegue nem explicar os dados de treino com alguma qualidade.

Não é dificil imaginar que é nesse meio termo que está o nosso modelo utilizável.

O overfitting é bem **mais comum que o underfitting**, mas vale o conhecimento sobre ambos.

Veja a imagem abaixo:

![over](https://experience.sap.com/wp-content/uploads/2018/02/illustrtaion-for-blog-06-700x300.jpg)


Não é, talvez, a melhor imagem para tratar do assunto, mas creio que dá para você compreender.

A figura **Fit** é aquela que consegue entender que há um padrão de dados e os consegue generalizar sem muita complexidade.

A **Undeffiting** não consegue separar os dados de nenhuma forma, é quase um algoritmo *lazy* ou *naive*, contudo a  culpa não é do algoritmo, mas da base de dados fraca.

**Overfitting** falar por si só, ele é muito complexo, por isso faz alguns "malabarismos" para classificar os dados.

* Aparentemente, esse modelo é o melhor para se utilizar, não?! Afinal, o que queremos é que um modelo acerte os dados. Mas vamos entender o porquê isso não é útil.
* Imagine que você tem dois grupos. Grupo **1.0** e grupo **2.0**; Se chegar um novo dado com um valor **1.0**, ele vai saber classificar como **1.0**, mas se chegar um valor **1.2**, por ele não saber generalizar, não vai saber onde colocar esse dados, por não ter aprendido a generalizar os dados, apenas trabalhar com os dados especificos.

## Checklist
1. Processos de Machine Learning
2. Teste de Overfitting (Você pode ir direto para esse ponto)

# 1. Processos de Machine Learning

**Importando a biblioteca pandas**

In [ ]:
import pandas as pd

**Carregando o dataset Spotify**

In [ ]:
dataset_spotify= pd.read_csv('/home/amador/dados/data.csv', sep=',')

**Separando os dados**

In [ ]:
classes = dataset_spotify['target']
dataset_spotify.drop('target', axis=1, inplace=True)

**Criando função pra remover coluna sempre que necessário**

In [ ]:
def remove_features(lista_features):
    for i in lista_features:
        dataset_spotify.drop(i, axis=1, inplace=True)
    return 'OK'

**Visualizando os dados**

In [ ]:
dataset_spotify.head(3)

id  acousticness  danceability  duration_ms  energy  instrumentalness  key  \
0   0        0.0102         0.833       204600   0.434          0.021900    2   
1   1        0.1990         0.743       326933   0.359          0.006110    1   
2   2        0.0344         0.838       185707   0.412          0.000234    2   

   liveness  loudness  mode  speechiness    tempo  time_signature  valence  \
0     0.165    -8.795     1       0.4310  150.062             4.0    0.286   
1     0.137   -10.401     1       0.0794  160.083             4.0    0.588   
2     0.159    -7.148     1       0.2890   75.044             4.0    0.173   

     song_title            artist  
0      Mask Off            Future  
1       Redbone  Childish Gambino  
2  Xanny Family            Future

**Removendo features não-representativas**

In [ ]:
remove_features(['id','song_title'])

'OK'

**Aplicando o get_dummies nos dados**

In [ ]:
dataset_spotify=pd.get_dummies(dataset_spotify, columns=['artist'], prefix=['artist'])

**Visualizando dados já com colunas removidas**

In [ ]:
dataset_spotify.head()

acousticness  danceability  duration_ms  energy  instrumentalness  key  \
0        0.0102         0.833       204600   0.434          0.021900    2   
1        0.1990         0.743       326933   0.359          0.006110    1   
2        0.0344         0.838       185707   0.412          0.000234    2   
3        0.6040         0.494       199413   0.338          0.510000    5   
4        0.1800         0.678       392893   0.561          0.512000    5   

   liveness  loudness  mode  speechiness  ...  artist_alt-J  artist_deadmau5  \
0    0.1650    -8.795     1       0.4310  ...             0                0   
1    0.1370   -10.401     1       0.0794  ...             0                0   
2    0.1590    -7.148     1       0.2890  ...             0                0   
3    0.0922   -15.236     1       0.0261  ...             0                0   
4    0.4390   -11.648     0       0.0694  ...             0                0   

   artist_for KING & COUNTRY  artist_one sonic society  artist_tUnE-yArDs  \
0                          0                         0                  0   
1                          0                         0                  0   
2                          0                         0                  0   
3                          0                         0                  0   
4                          0                         0                  0   

   artist_tobyMac  artist_권나무 Kwon Tree  artist_도시총각 Dosichonggak  \
0               0                     0                         0   
1               0                     0                         0   
2               0                     0                         0   
3               0                     0                         0   
4               0                     0                         0   

   artist_카우칩스 The CowChips  artist_플랫핏 Flat Feet  
0                         0                     0  
1                         0                     0  
2                         0                     0  
3                         0                     0  
4                         0                     0  

[5 rows x 1356 columns]

**Dividindo os dados em treino e teste em 80/20**

In [ ]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test=train_test_split(dataset_spotify,classes,test_size=0.2,random_state=123)

**Visualizando os dados de treino com as features**

In [ ]:
X_train

acousticness  danceability  duration_ms  energy  instrumentalness  key  \
1505      0.004770         0.585       214740   0.614          0.000155   10   
813       0.114000         0.665       216728   0.513          0.303000    0   
615       0.030200         0.798       216585   0.481          0.000000    7   
319       0.106000         0.912       194040   0.317          0.000208    6   
320       0.021100         0.697       236456   0.905          0.893000    6   
...            ...           ...          ...     ...               ...  ...   
2012      0.001060         0.584       274404   0.932          0.002690    1   
1346      0.000021         0.535       203500   0.974          0.000149   10   
1406      0.503000         0.410       256333   0.648          0.000000    7   
1389      0.705000         0.894       222307   0.161          0.003300    4   
1534      0.623000         0.470       394920   0.156          0.187000    2   

      liveness  loudness  mode  speechiness  ...  artist_alt-J  \
1505    0.0762    -5.594     0       0.0370  ...             0   
813     0.1220    -7.314     1       0.3310  ...             0   
615     0.1280   -10.488     1       0.3140  ...             0   
319     0.0723   -12.719     0       0.0378  ...             0   
320     0.1190    -7.787     0       0.0339  ...             0   
...        ...       ...   ...          ...  ...           ...   
2012    0.1290    -3.501     1       0.3330  ...             0   
1346    0.2630    -3.566     0       0.1720  ...             0   
1406    0.2190    -4.469     1       0.0362  ...             0   
1389    0.3120   -14.311     1       0.0880  ...             0   
1534    0.1040   -17.036     1       0.0399  ...             0   

      artist_deadmau5  artist_for KING & COUNTRY  artist_one sonic society  \
1505                0                          0                         0   
813                 0                          0                         0   
615                 0                          0                         0   
319                 0                          0                         0   
320                 0                          0                         0   
...               ...                        ...                       ...   
2012                0                          0                         0   
1346                0                          0                         0   
1406                0                          0                         0   
1389                0                          0                         0   
1534                0                          0                         0   

      artist_tUnE-yArDs  artist_tobyMac  artist_권나무 Kwon Tree  \
1505                  0               0                     0   
813                   0               0                     0   
615                   0               0                     0   
319                   0               0                     0   
320                   0               0                     0   
...                 ...             ...                   ...   
2012                  0               0                     0   
1346                  0               0                     0   
1406                  0               0                     0   
1389                  0               0                     0   
1534                  0               0                     0   

      artist_도시총각 Dosichonggak  artist_카우칩스 The CowChips  artist_플랫핏 Flat Feet  
1505                         0                         0                     0  
813                          0                         0                     0  
615                          0                         0                     0  
319                          0                         0                     0  
320                          0                         0                     0  
...                        ...                  

**Visualizando os dados de treino com a classe**

In [ ]:
y_train

1505    0
813     1
615     1
319     1
320     1
       ..
2012    0
1346    0
1406    0
1389    0
1534    0
Name: target, Length: 1613, dtype: int64

**Visualizando os dados de teste com features**

In [ ]:
X_test

acousticness  danceability  duration_ms  energy  instrumentalness  key  \
1423       0.00184         0.541       225427   0.609          0.000000    1   
356        0.00855         0.293       202160   0.792          0.011100    7   
854        0.93400         0.440       254360   0.433          0.811000    2   
1445       0.09870         0.443       210884   0.727          0.000009    0   
1647       0.00346         0.723       176561   0.809          0.001230    7   
...            ...           ...          ...     ...               ...  ...   
156        0.00212         0.777       500550   0.739          0.868000    9   
883        0.19800         0.890       326318   0.520          0.273000    2   
853        0.27100         0.567       336800   0.851          0.000000    1   
903        0.02450         0.803       316373   0.427          0.405000    8   
890        0.03280         0.598       260400   0.657          0.000027    8   

      liveness  loudness  mode  speechiness  ...  artist_alt-J  \
1423    0.1790    -3.715     1       0.1270  ...             0   
356     0.1330    -3.747     1       0.0363  ...             0   
854     0.2730   -17.453     0       0.0412  ...             0   
1445    0.1180    -5.171     0       0.0984  ...             0   
1647    0.5650    -3.081     0       0.0625  ...             0   
...        ...       ...   ...          ...  ...           ...   
156     0.1090    -9.089     1       0.0390  ...             0   
883     0.0638    -6.927     0       0.1870  ...             0   
853     0.5690    -5.897     1       0.4840  ...             0   
903     0.0953   -10.819     1       0.0509  ...             0   
890     0.1720    -7.466     1       0.0386  ...             0   

      artist_deadmau5  artist_for KING & COUNTRY  artist_one sonic society  \
1423                0                          0                         0   
356                 0                          0                         0   
854                 0                          0                         0   
1445                0                          0                         0   
1647                0                          0                         0   
...               ...                        ...                       ...   
156                 0                          0                         0   
883                 0                          0                         0   
853                 0                          0                         0   
903                 0                          0                         0   
890                 0                          0                         0   

      artist_tUnE-yArDs  artist_tobyMac  artist_권나무 Kwon Tree  \
1423                  0               0                     0   
356                   0               0                     0   
854                   0               0                     0   
1445                  0               0                     0   
1647                  0               0                     0   
...                 ...             ...                   ...   
156                   0               0                     0   
883                   0               0                     0   
853                   0               0                     0   
903                   0               0                     0   
890                   0               0                     0   

      artist_도시총각 Dosichonggak  artist_카우칩스 The CowChips  artist_플랫핏 Flat Feet  
1423                         0                         0                     0  
356                          0                         0                     0  
854                          0                         0                     0  
1445                         0                         0                     0  
1647                         0                         0                     0  
...                        ...                  

**Visualizando os dados de teste com a classe**

In [ ]:
y_test

1423    0
356     1
854     1
1445    0
1647    0
       ..
156     1
883     1
853     1
903     1
890     1
Name: target, Length: 404, dtype: int64

**Treinando o classificador clf com SVM**

In [ ]:
# importação do algoritmo
from sklearn import svm
# instanciação do algoritmo
clf = svm.SVC().fit(dataset_spotify,classes)
# treinando o algoritmo
clf.fit(X_train,y_train)

SVC()

# 2. Teste de Overfitting

**Importando bibliotecas**

In [ ]:
from sklearn.metrics import accuracy_score
from sklearn import metrics

**Predições de resultados com dados de teste**

In [ ]:
y_pred=clf.predict(X_test)

**Imprimindo acurácia com predições e dados de teste**

In [ ]:
metrics.accuracy_score(y_test,y_pred)

0.5668316831683168

**Predições de resultados com dados de treino**

In [ ]:
y_pred2=clf.predict(X_train)

**Imprimindo acurácia com predições e dados de treino**

In [ ]:
metrics.accuracy_score(y_train,y_pred2)

0.5523868567885927

### Houve Overfitting?

Nesse caso, não. Veja que os **resultados foram bem próximos** (56% e 55%). 

Se, ao alterar os parâmetros, os dados de treino começarem a melhorar o percentual da acurácia e os de teste continuarem estagnados ou não forem subindo na mesma, ou quase na mesma proporção, o modelo estará começando a sofrer overfitting.

O **Cientista de Dados** tem que observar e avaliar sempre!

# Obrigado!

Obrigado por ter disponibilizado um pouco do seu tempo e atenção aqui. Espero que, de alguma forma, tenha sido útil para seu crescimento. Se houver qualquer dúvida ou sugestão, não hesite em entrar em contato no [LinkedIn]((https://www.linkedin.com/in/daniel-sousa-amador)) e verificar meus outros projetos no [GitHub](https://github.com/danielamador12/Portfolio/blob/master/README.md).

[![LinkedIn](https://img.shields.io/badge/LinkedIn-DanielSousaAmador-purple.svg)](https://www.linkedin.com/in/daniel-sousa-amador)
[![GitHub](https://img.shields.io/badge/GitHub-danielamador12-darkblue.svg)](https://github.com/danielamador12)
[![Medium](https://img.shields.io/badge/Medium-DanielSousaAmador-darkorange.svg)](https://medium.com/@daniel.s.amador)



<center><img width="90%" src="https://raw.githubusercontent.com/danielamador12/Portfolio/master/github.png"></center>